### ISSUE TITLE
[Question] Topics getting deleted on reinstall of Kafka with persistent storage

### ISSUE THREAD

**Issue by Ihjaz** [strimzi/strimzi-kafka-operator#1593]
Hi,

I've using standalone topic operator. I have enabled persistence for Kafka.

I install the helm release and create topics using topic-operator and then delete the helm release and then I reinstall the helm release.

What I see is that topic operator is deleting the topics from the first installation.

```
2019-04-29 10:26:03 INFO  TopicOperator:376 - Reconciling topic test2, k8sTopic:null, kafkaTopic:nonnull, privateTopic:nonnull
2019-04-29 10:26:03 DEBUG TopicOperator:419 - KafkaTopic deleted in k8s => delete topic from kafka and from topicStore
2019-04-29 10:26:03 DEBUG TopicOperator:561 - Enqueuing event DeleteKafkaTopic(topicName=test2)
2019-04-29 10:26:03 INFO  TopicOperator:376 - Reconciling topic test, k8sTopic:null, kafkaTopic:nonnull, privateTopic:nonnull
2019-04-29 10:26:03 DEBUG TopicOperator:419 - KafkaTopic deleted in k8s => delete topic from kafka and from topicStore
2019-04-29 10:26:03 DEBUG TopicOperator:561 - Enqueuing event DeleteKafkaTopic(topicName=test)
2019-04-29 10:26:03 INFO  TopicOperator:321 - Deleting topic 'test2'
2019-04-29 10:26:03 DEBUG BaseKafkaImpl:217 - Deleting topic test2
```

I want to reuse the topics from the first installation and it should be available after the reinstall. 

Is the behavior that I'm seeing expected or is there something I need to do differently to prevent the operator from deleting the topics?

**Comment by tombentley** [comment]
I suspect what is happening is that by deleting the helm release you're deleting the CRDs (including the `KafkaTopic` CRD), which will cause Kubernetes to delete the `KafkaTopic` resources. If the Topic Operator is running perhaps it observes that deletion and so deletes the Kafka topic. (Maybe it then starts getting errors until the `KafkaTopic` CRD is reinstalled, but by then the damage has been done). 

Have you tried uninstalling/scaling to zero the standalone TO before deleting the helm release, and only after reinstalling via helm restarting the TO?

**Comment by Ihjaz** [comment]
Hi,

I have the Topic operator also in the helm chart. It is basically the Confluent's cp-helm-charts ( https://github.com/confluentinc/cp-helm-charts) to which I've added Topic operator also.

So TO is getting installed and uninstalled as part of the helm release.

If you notice the log above, when the helm chart is reinstalled it is finding the object k8sTopic to be null and hence deletes the topics. 

Do you have any idea why k8sTopic would be null here?

**Comment by scholzj** [comment]
Right. I think the idea suggested by Tom is that when you uninstall the Helm Chart, you delete the custom resource definition for the `KafkaTopic` resource. And that will also delete all instances of it. So if the reinstallation is in reality really delete everything and create from scratch, it would end up this way. You would probably need to just apply the changes instead of deleting and creating to avoid this. But I'm not really a Helm expert, so I'm not sure how to do it.

**Comment by tombentley** [comment]
There's another thing which could be the cause, I think. The Topic Operator maintains a state store (under the `strimzi` znode) in zookeeper about each topic. That won't be deleted when you uninstall the Topic Operator, which will affect the algorithm when you start reinstall the operator. It's also worth mentioning that there are some algorithm improvements on master which might be beneficial.

**Comment by Ihjaz** [comment]
Hi,

Could you tell me what the following variables in the above logs represent ?

k8sTopic
kafkaTopic
privateTopic

**Comment by tombentley** [comment]
* k8sTopic – the `KafkaTopic` in Kubernetes/OpenShift.
* kafkaTopic – the topic in Kafka.
* privateTopic – a copy of the topic state held under the `/strimzi` znode in Zookeeper

**Comment by Ihjaz** [comment]
Thanks @tombentley 

So with persistence enabled, when I reinstall the chart I guess the topic state is retained in the zookeeper from persistence and the topic is also there in Kafka from persistence making "kafkaTopic" and "privateTopic" non-null but there is no corresponding KafkaTopic in the Kubernetes.

And the current implementation considers the state **k8sTopic:null, kafkaTopic:nonnull, privateTopic:nonnull** as topic is deleted and hence deletes the topics from Kafka and private store.

**Comment by tombentley** [comment]
What version of Strimzi are you using?

**Comment by Ihjaz** [comment]
I'm using Release 0.11.1 .

**Comment by tombentley** [comment]
9f42a003dcb4b3fdcaa90484495ac25abbef39f0 made some fixes to the algorithm the TO applies at start up. However, I don't think that will fix this case on its own because it still appears that the k8s topic was deleted since the last reconciliation. I think you will also have to manually delete the `/strimzi` subtree in zookeeper while the TO is not running (e.g. scaled to zero), so it appears to the TO that the topics have been created on the Kafka side and need creating on the Kube side.

9f42a003dcb4b3fdcaa90484495ac25abbef39f0 is only on master at the moment. @scholzj I thought we should have cherry-picked 9f42a003dcb4b3fdcaa90484495ac25abbef39f0 into 0.11.3, no?

**Comment by Ihjaz** [comment]
Manually deleting the /strimzi data prevents the TO from deleting the topics on reinstall but it is not able to successfully reconcile the topics from kafka. 

```
2019-05-15 13:31:01 DEBUG TopicOperator:1006 - Reconciling kafka topics [eventing-kafka-cp-kafka-connect-config, test2, test, _schemas, eventing-kafka-cp-kafka-connect-offset, test1, eventing-kafka-cp-kafka-connect-status]
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic eventing-kafka-cp-kafka-connect-config
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic test2
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic test
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic _schemas
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic eventing-kafka-cp-kafka-connect-offset
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic test1
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 20,4  replyHeader:: 20,25769804031,-101  request:: '/strimzi/topics/eventing-kafka-cp-kafka-connect-config,F  response::
2019-05-15 13:31:01 DEBUG TopicOperator:1009 - periodic reconciliation of topic eventing-kafka-cp-kafka-connect-status
2019-05-15 13:31:01 DEBUG BaseKafkaImpl:94 - Handler for work listTopics487995684 executed ok
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 21,4  replyHeader:: 21,25769804031,-101  request:: '/strimzi/topics/test2,F  response::
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 22,4  replyHeader:: 22,25769804031,-101  request:: '/strimzi/topics/test,F  response::
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 23,4  replyHeader:: 23,25769804031,-101  request:: '/strimzi/topics/_schemas,F  response::
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 24,4  replyHeader:: 24,25769804031,-101  request:: '/strimzi/topics/eventing-kafka-cp-kafka-connect-offset,F  response::
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 25,4  replyHeader:: 25,25769804031,-101  request:: '/strimzi/topics/test1,F  response::
2019-05-15 13:31:01 DEBUG ClientCnxn:846 - Reading reply sessionid:0x10004befd670004, packet:: clientPath:null serverPath:null finished:false header:: 26,4  replyHeader:: 26,25769804031,-101  request:: '/strimzi/topics/eventing-kafka-cp-kafka-connect-status,F  response::
```

It is not showing the topics in both kubernetes and /strimzi 

```
]# kubectl get kafkatopic 
No resources found.
```

```
bash-4.2# zookeeper-shell.sh 127.0.0.1:2181
Connecting to 127.0.0.1:2181
Welcome to ZooKeeper!

ls /strimzi/topics
[]
```

**Comment by tombentley** [comment]
@Ihjaz was that using master or 0.11.1 still? You should be using master.

**Comment by Ihjaz** [comment]
I was using 0.11.1 .

**Comment by tombentley** [comment]
Yes, I think you should try it with master.

**Comment by Ihjaz** [comment]
Hi All,

I upgraded my code from 0.11.1 to 0.11.4 to pick the new changes for reconcile.

I then changed the create mode from PERSISTENCE to EPHEMERAL in the create() method in ZkTopicStore. 
So now whenever I delete the helm chart, the topics under /strimzi/topics on znode gets deleted as it is ephemeral and when the new chart is installed both the k8sTopic and privateTopic is null and it does the reconciliation successfully.

Do you guys forsee any issues in changing the create mode from PERSISTENCE to EPHEMERAL ?

**Comment by tombentley** [comment]
I would recommend not to do that. An ephemeral znode will only exist while the creating process remains connected to the zookeeper cluster. So if the pod get restarted or there are transient network problems the znode will be deleted, as you say. The Topic Operator was written with the assumption that the topic store was durable. There are no tests for what happens if the topic store gets deleted (either while the TO is down or while it is running), so really I could not be confident that it would work property in all scenarios. 

For example, I expect that if a topic is deleted in Kafka or Kubernetes while the TO is down then the deletion would not be reflected on TO startup, but rather the topic would be recreated. Similarly if a topic is modified while the TO is not running then it won't know which end changed while it was down and so the changes could get reverted.

**Comment by Ihjaz** [comment]
Hi,

I solved this issue by having a 'pre-delete' helm hook which will delete the /strimzi/topics from the zookeeper.

**Comment by scholzj** [comment]
I will close this since there was no update for more than 10 days. If you have something more feel free to open a new issue, reopen this one or get in touch with us on Slack or mailing list.

BTW: The process of recovering cluster from PVs or PVCs is now also described in the docs: https://strimzi.io/docs/latest/full.html#cluster-recovery_str
